engine/execution/ingestion/stop_control.go

```diff
@@ -0,0 +1,235 @@
+		return oldHeight, oldCrash, fmt.Errorf("cannot update stop height, already paused")
+	}
+
+	// +1 because we track last executing height, so +1 is the lowest possible block to stop
+	if height <= s.lastExecutingHeight+1 {
+		return oldHeight, oldCrash, fmt.Errorf("cannot update stop height, given height %d at or below last executed %d", height, s.lastExecutingHeight)
+	}
+
+	s.log.Info().
+		Uint64("height", height).Bool("crash", crash).
+		Uint64("old_height", oldHeight).Bool("old_crash", oldCrash).Msg("new stop height set")
+
+	s.log.Debug().Int8("previous_state", int8(s.state)).Int8("new_state", int8(StopControlSet)).Msg("StopControl state transition")
+	s.state = StopControlSet
+
+	s.height = height
+	s.crash = crash
+	s.stopAfterExecuting = flow.ZeroID
+
+	return oldHeight, oldCrash, nil
+}
+
+// GetStopHeight returns:
+//   - height
+//   - crash
+//
+// Values are undefined if they were not previously set
+func (s *StopControl) GetStopHeight() (uint64, bool) {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.height, s.crash
+}
+
+// blockProcessable should be called when new block is processable.
+// It returns boolean indicating if the block should be processed.
+func (s *StopControl) blockProcessable(b *flow.Header) bool {
+
+	s.Lock()
+	defer s.Unlock()
+
+	if s.state == StopControlOff {
+		return true
+	}
+
+	if s.state == StopControlPaused {
+		return false
+	}
+
+	// skips blocks at or above requested stop height
+	if b.Height >= s.height {
+		s.log.Warn().Msgf("Skipping execution of %s at height %d because stop has been requested at height %d", b.ID(), b.Height, s.height)
+		s.log.Debug().Int8("previous_state", int8(s.state)).Int8("new_state", int8(StopControlCommenced)).Msg("StopControl state transition")
+		s.state = StopControlCommenced // if block was skipped, move into commenced state
+		return false
+	}
+
+	return true
+}
+
+// blockFinalized should be called when a block is marked as finalized
+func (s *StopControl) blockFinalized(ctx context.Context, execState state.ReadOnlyExecutionState, h *flow.Header) {
+
+	s.Lock()
+	defer s.Unlock()
+
+	if s.state == StopControlOff || s.state == StopControlPaused {
+		return
+	}
+
+	// Once finalization reached stop height we can be sure no other fork will be valid at this height,
+	// if this block's parent has been executed, we are safe to stop or crash.
+	// This will happen during normal execution, where blocks are executed before they are finalized.
+	// However, it is possible that EN block computation progress can fall behind. In this case,
+	// we want to crash only after the execution reached the stop height.
+	if h.Height == s.height {
+
+		executed, err := state.IsBlockExecuted(ctx, execState, h.ParentID)
+		if err != nil {
+			// any error here would indicate unexpected storage error, so we crash the node
+			s.log.Fatal().Err(err).Str("block_id", h.ID().String()).Msg("failed to check if the block has been executed")
+			return
+		}
+
+		if executed {
+			s.stopExecution()
```

Comment:
optional: wonder if would be useful to pass in the finalized block ID for it to be logged.

Reply:
Yep, won't hurt